lemur/plugins/lemur_acme/ultradns.py

```diff
@@ -0,0 +1,323 @@
+    access_token = get_ultradns_token()
+    return {"Authorization": "Bearer {}".format(access_token), "Content-Type": "application/json"}
+
+
+def _paginate(path, key):
+    limit = 100
+    params = {"offset": 0, "limit": 1}
+    resp = _get(path, params)
+    for index in range(0, resp["resultInfo"]["totalCount"], limit):
+        params["offset"] = index
+        params["limit"] = limit
+        resp = _get(path, params)
+        yield resp[key]
+
+
+def _get(path, params=None):
+    # Function to execute a GET request on the given URL (base_uri + path) with given params
+    base_uri = current_app.config.get("ACME_ULTRADNS_DOMAIN", "")
+    resp = requests.get(
+        "{0}{1}".format(base_uri, path),
+        headers=_generate_header(),
+        params=params,
+        verify=True,
+    )
+    resp.raise_for_status()
+    return resp.json()
+
+
+def _delete(path):
+    # Function to execute a DELETE request on the given URL
+    base_uri = current_app.config.get("ACME_ULTRADNS_DOMAIN", "")
+    resp = requests.delete(
+        "{0}{1}".format(base_uri, path),
+        headers=_generate_header(),
+        verify=True,
+    )
+    resp.raise_for_status()
+
+
+def _post(path, params):
+    # Executes a POST request on given URL. Body is sent in JSON format
+    base_uri = current_app.config.get("ACME_ULTRADNS_DOMAIN", "")
+    resp = requests.post(
+        "{0}{1}".format(base_uri, path),
+        headers=_generate_header(),
+        data=json.dumps(params),
+        verify=True,
+    )
+    resp.raise_for_status()
+
+
+def _has_dns_propagated(name, token, domain):
+    # Check whether the DNS change made by Lemur have propagated to the public DNS or not.
+    # Invoked by wait_for_dns_change() function
+    txt_records = []
+    try:
+        dns_resolver = dns.resolver.Resolver()
+        dns_resolver.nameservers = [domain]
+        dns_response = dns_resolver.query(name, "TXT")
+        for rdata in dns_response:
+            for txt_record in rdata.strings:
+                txt_records.append(txt_record.decode("utf-8"))
+    except dns.exception.DNSException:
+        metrics.send("has_dns_propagated_fail", "counter", 1)
+        return False
+
+    for txt_record in txt_records:
+        if txt_record == token:
+            metrics.send("has_dns_propagated_success", "counter", 1)
+            return True
+
+    return False
+
+
+def wait_for_dns_change(change_id, account_number=None):
+    # Waits and checks if the DNS changes have propagated or not.
+    # First check the domains authoritative server. Once this succeeds,
+    # we ask a public DNS server (Google <8.8.8.8> in our case).
+    fqdn, token = change_id
+    number_of_attempts = 20
+    nameserver = get_authoritative_nameserver(fqdn)
+    for attempts in range(0, number_of_attempts):
+        status = _has_dns_propagated(fqdn, token, nameserver)
+        current_app.logger.debug("Record status on ultraDNS authoritative server for fqdn: {}: {}".format(fqdn, status))
```

Comment:
Can you log in dictionary format for all the places you are logging? This will (eventually) be converted to JSON so we can parse out fields from the log better.
Example:
function = f"{__name__}.{sys._getframe().f_code.co_name}"
log_data = {
    "function": function,
    "fqdn": fqdn,
    "status": status,
    "message": "Record status on ultraDNS authoratative server"
}
current_app.logger.debug(log_data)